src/Middleware/RequestDecompression/src/SizeLimitedStream.cs

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.RequestDecompression;
+
+internal class SizeLimitedStream : Stream
+{
+    private readonly Stream _innerStream;
+    private readonly long? _sizeLimit;
+
+    private long _totalBytesRead;
+
+    public SizeLimitedStream(Stream innerStream, long? sizeLimit)
```

Comment:
What do known encodings have to do with it? We should always be enforcing a limit on the decompressed request body.

Reply:
[Here ](https://github.com/dotnet/aspnetcore/pull/40279#discussion_r810507794) I was saying that for known encodings, instead of creating an instance of `SizedLimitStream` and pass an instance of `GZipStream` in its ctor, we write a new class called `SizedLimitGZipStream` which inherits `GZipStream` and check the size on its Read method. This way instead of allocating two instances of `GZipSteam` and `SizedLimitStream` we only allocate one `SizedLimitGZipStream`. 

The _known encoding_ part is for that we don't know if the stream returning from the provider is checking the size, so (here comes the ugly part!) `SizedLimitGZipStream` should also implement a marker interface, so the middleware only wrap the stream in `SizedLimitStream` if the stream is not implementing the marker interface. The marker interface could be internal of course.

Reply:
That's going to be a lot of extra code to save that allocation. Let's not for now.

Reply:
I blame @davidfowl for making me think this way. I've got _allocaphobia_ (fear of allocation). 😅

Reply:
![JackNicholsonYesGIF](https://user-images.githubusercontent.com/95136/156197037-488dafea-a627-4442-8ff2-d5adadcacc6d.gif)

Reply:
We can optimize after the feature works if the gains show its worth it. I'm all about not allocating per request.